cargo-miri/bin.rs

```diff
@@ -627,15 +656,55 @@ fn phase_cargo_rustc(args: env::Args) {
         // like we want them.
         // Instead of compiling, we write JSON into the output file with all the relevant command-line flags
         // and environment variables; this is used when cargo calls us again in the CARGO_TARGET_RUNNER phase.
-        store_json(CrateRunInfo::collect(args));
+        let info = CrateRunInfo::collect(args);
+        store_json(&info);
+
+        // Rustdoc expects us to exit with an error code if the test is marked as `compile_fail`,
+        // just creating the JSON file is not enough: we need to detect syntax errors,
+        // so we need to run Miri with `MIRI_BE_RUSTC` for a check-only build.
+        if std::env::var_os("MIRI_CALLED_FROM_RUSTDOC").is_some() {
+            let mut cmd = miri();
+            let env = if let CrateRunInfo::RunWith(env) = info {
+                env
+            } else {
+                return;
```

Comment:
Can this ever happen? If not, it might be better to make this `unreachable!()`, just to be sure.